### PR TITLE
Don't strip hyphen from BasePathMapping

### PR DESCRIPTION
**Pull request by GavinZZ**
### Issue #, if available
https://github.com/aws/serverless-application-model/issues/2279

### Description of changes
Input template 
```
Globals:
  Api:
    Domain:
      BasePath: foo-bar
```

Observed output template
```
"MyApifoobarBasePathMapping": {
  "Type": "AWS::ApiGateway::BasePathMapping",
  "Properties": {
    "BasePath": "foobar"
  }
}
```

Expected output template
```
"MyApifoobarBasePathMapping": {
  "Type": "AWS::ApiGateway::BasePathMapping",
  "Properties": {
    "BasePath": "foo-bar"
  }
}
```

### Checklist

- [ ] Adheres to the [development guidelines](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#development-guidelines)
- [ ] Add/update [unit tests](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#unit-testing-with-multiple-python-versions) using:
    - [ ] Correct values
    - [ ] Bad/wrong values (None, empty, wrong type, length, etc.)
    - [ ] [Intrinsic Functions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html)
- [ ] Add/update [integration tests](https://github.com/aws/serverless-application-model/blob/develop/INTEGRATION_TESTS.md)
- [ ] Update documentation
- [ ] Verify transformed template deploys and application functions as expected
- [ ] Do these changes include any template validations?
    - [ ] Did the newly validated properties support intrinsics prior to adding the validations? (If unsure, please review [Intrinsic Functions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html) before proceeding).
        - [ ] Does the pull request ensure that intrinsics remain functional with the new validations?

### Examples?

Please reach out in the comments, if you want to add an example. Examples will be 
added to `sam init` through https://github.com/awslabs/aws-sam-cli-app-templates/

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samtranslator/model/api/api_generator.py [hunk 1]
```diff
@@ -513,16 +513,16 @@ def _construct_api_domain(self, rest_api, route53_record_set_groups):  # type: i
             basepath_mapping.Stage = ref(rest_api.logical_id + ".Stage")
             basepath_resource_list.extend([basepath_mapping])
         else:
-            for path in basepaths:
-                path = "".join(e for e in path if e.isalnum())
+            for basepath in basepaths:
+                path = "".join(e for e in basepath if e.isalnum())
```
Comment (hoffa) [inline samtranslator/model/api/api_generator.py:517]: What changed? Any tests?

Comment (GavinZZ) [reply samtranslator/model/api/api_generator.py:517]: oh wait, the test file changes didn't got committed. Let me add that

File: tests/translator/output/api_with_basic_custom_domain.json [hunk 2]
```diff
@@ -165,7 +165,7 @@
     },
     "MyApifetchBasePathMapping": {
       "Properties": {
-        "BasePath": "fetch",
+        "BasePath": "/fetch",
```
Comment (hoffa) [inline tests/translator/output/api_with_basic_custom_domain.json:168]: Test outputs should typically never change. Why are these changing?

Comment (GavinZZ) [reply tests/translator/output/api_with_basic_custom_domain.json]: removed. Changed the code to match `Http API` where we remove trailing and leading `/`.

File: tests/translator/input/api_with_custom_base_path.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+Globals:
+  Api:
+    Domain:
+      BasePath: foo-bar
```
Comment (hoffa) [inline tests/translator/input/api_with_custom_base_path.yaml:4]: I don't see this in the output files. 🤔

Comment (hoffa) [reply tests/translator/input/api_with_custom_base_path.yaml:4]: And how is this relevant to https://github.com/aws/serverless-application-model/issues/2279#issuecomment-1007055489?

Does something change with `Globals` versus `Properties`?

File: samtranslator/model/api/api_generator.py [hunk 4]
```diff
@@ -513,16 +513,18 @@ def _construct_api_domain(self, rest_api, route53_record_set_groups):  # type: i
             basepath_mapping.Stage = ref(rest_api.logical_id + ".Stage")
             basepath_resource_list.extend([basepath_mapping])
         else:
-            for path in basepaths:
-                path = "".join(e for e in path if e.isalnum())
+            for basepath in basepaths:
+                # ignore leading and trailing `/` in the path name
```
Comment (hoffa) [inline samtranslator/model/api/api_generator.py:517]: Why?

File: samtranslator/model/api/api_generator.py [hunk 5]
```diff
@@ -513,16 +513,18 @@ def _construct_api_domain(self, rest_api, route53_record_set_groups):  # type: i
...
+                basepath = basepath.strip("/")
+                path = "".join(e for e in basepath if e.isalnum())
```
[the remark this answers is not in the document]
Comment (GavinZZ) [reply samtranslator/model/api/api_generator.py:519]: path is for the logical id.  basepath is used for `BasePath` property.

[the remark this answers is not in the document]
Comment (aaythapa) [reply samtranslator/model/api/api_generator.py:519]: ah i missed that 👍

**Review by aaythapa — APPROVED** [review]

File: samtranslator/model/api/api_generator.py (hunk 4 above)
Comment (GavinZZ) [reply samtranslator/model/api/api_generator.py:517]: `A base path may only contain letters, numbers, and one of "$-_.+!*'()"`

Comment (hoffa) [reply samtranslator/model/api/api_generator.py:517]: That doesn't explain why it's only stripping slashes.

File: samtranslator/model/api/api_generator.py [hunk 6]
```diff
@@ -502,6 +502,9 @@ def _construct_api_domain(self, rest_api, route53_record_set_groups):  # type: i
         else:
             basepaths = None
 
+        # Boolean to allow/disallow hyphens in BasePath property
+        strip_hyphens = self.domain.get("StripHyphens", True)
```
Comment (hoffa) [inline samtranslator/model/api/api_generator.py:506]: For future, we should have some way of specifying the default in the props definition, so we don't need to sprinkle custom logic like this.

File: samtranslator/model/api/api_generator.py [hunk 7]
```diff
@@ -513,16 +516,19 @@ def _construct_api_domain(self, rest_api, route53_record_set_groups):  # type: i
             basepath_mapping.Stage = ref(rest_api.logical_id + ".Stage")
             basepath_resource_list.extend([basepath_mapping])
         else:
-            for path in basepaths:
-                path = "".join(e for e in path if e.isalnum())
+            for basepath in basepaths:
+                # Remove possible leading and trailing '/' because a base path may only
+                # contain letters, numbers, and one of "$-_.+!*'()"
+                basepath = basepath.strip("/")
```
Comment (hoffa) [inline samtranslator/model/api/api_generator.py:522]: The `strip_hyphens` ternary would be much clearer here, as this is dead code if it's `True`.

File: tests/translator/output/api_with_custom_base_path.json [hunk 8]
```diff
@@ -99,6 +99,57 @@
...
+        "RestApiId": {
+          "Ref": "MyApiWithStripHyphens"
+        },
+        "StageName": "Prod"
+      },
+      "Type": "AWS::ApiGateway::Stage"
+    },
+    "MyApiWithStripHyphensfoobarBasePathMapping": {
+      "Properties": {
+        "BasePath": "foo-bar",
```
Comment (hoffa) [inline tests/translator/output/api_with_custom_base_path.json:140]: `WithStripHyphens` but still has hyphens?

**Review by xazhao — APPROVED** [review]

File: samtranslator/model/api/api_generator.py (hunk 6 above)
Comment (GavinZZ) [reply samtranslator/model/api/api_generator.py:506]: Agree.

File: tests/translator/input/api_with_custom_base_path.yaml [hunk 9]
```diff
@@ -32,3 +32,18 @@ Resources:
...
+  MyApiWithoutStripHyphens:
+    Type: AWS::Serverless::Api
+    Properties:
+      OpenApiVersion: 3.0.1
+      StageName: Prod
+      Domain:
+        DomainName: MyEdgeDomainName
+        CertificateArn: MyEdgeDomainCert
+        EndpointConfiguration: EDGE
+        StripHyphens: false
```
Comment (hoffa) [inline tests/translator/input/api_with_custom_base_path.yaml:45]: 1. Is `StripHyphens` clear enough from context? An uglier but clearer alternative could be `BasePathStripHyphens`.
2. Is `StripHyphens` correct? I don't see code specifically stripping hyphens, but anything non-alphanumeric.

Comment (GavinZZ) [reply tests/translator/input/api_with_custom_base_path.yaml:45]: Good call. `BasePathStripSymbols` is probably better in this context.

Comment (hoffa) [reply tests/translator/input/api_with_custom_base_path.yaml]: Symbol is quite a fancy word, and makes me think of symbol tables in binaries. Plus any character can really be a "symbol".

Some bad alternatives:

- `BasePathSanitize`/`SanitizeBasePath`
- `BasePathAlphanumeric`

**Review by aahung — APPROVED** [review]

File: tests/translator/input/api_with_custom_base_path.yaml [hunk 10]
```diff
@@ -32,3 +32,18 @@ Resources:
...
+  MyApiWithoutNormalizedBasePath:
+    Type: AWS::Serverless::Api
+    Properties:
+      OpenApiVersion: 3.0.1
+      StageName: Prod
+      Domain:
+        DomainName: MyEdgeDomainName
+        CertificateArn: MyEdgeDomainCert
+        EndpointConfiguration: EDGE
+        NormalizeBasePath: false
```
Comment (hoffa) [inline tests/translator/input/api_with_custom_base_path.yaml:45]: Would be clearer if it used `BasePath` with slashes as well, to make sure we don't revert this later and tests keep passing.

**Review by hoffa — APPROVED** [review]

**Pull request merged**